feat(squash): Add support for erofs

prudo1 opened the pull request:
This pull request adds support for erofs for the 99squash module. For this it first cleans up the special handling needed in dracut.sh, then splits out the squashfs specific code into 95squash-squashfs and, finally adds a new 95squash-erofs module. The "new" 99squash module is designed in a way that it picks an appropriate back end automatically if added directly and none of the back ends is provided. Default is to use squashfs to keep the current behavior.

There are still some small things to do in lsinitrd and dracut-rescue-initramfs. Nevertheless, it would be good to get some feedback about the overall approach.

pvalena commented on lsinitrd.sh:
```diff
@@ -222,17 +222,30 @@
+	erofs-root.img)
+		mkdir -p "$EROFS_TMPMNT"
+		mount -t erofs "$SQUASH_TMPFILE" "$EROFS_TMPMNT"
+		(cd $EROFS_TMPMNT; find "$EROFS_TMPMNT" -ls)
```
https://github.com/redhat-plumbers/dracut-rhel10/pull/7/commits/6bcfe67d6b19d303cb1c4871c04ef18c74f7550f

Also, I've just realized you are be using tabs instead of spaces :) please use spaces in dracut code.

prudo1 replied:
Yeah, I saw the shfmt errors as well. But then it was after 6pm on a Friday and I decided that it's something that can wait till Monday ;-)

pvalena reviewed (changes requested):
Ad `Commisery / Conventional Commit Message Checker` you need to start the message with lowercase letter.

Otherwise LGTM!

LaszloGombos commented:
Thank you @prudo1 for contributing this new feature and improving and testing `squash` module.

One observation I have (independently from the `erofs` feature addition) that the new `squash_install` functions (old `installpost` function) make certain assumptions about `busybox compilation`. It assumes that busybox provides `switch_root` as an example.

`busybox` can laso be installed using the `busybox` dracut module, which has support to mix `busybox` applets and e.g. `coreutils` implementations when needed.

I was wondering if you would consider installing `busybox` calling `module_install` (similarly as `squash_get_handler`) gets installed.

prudo1 commented:
> Thank you @prudo1 for contributing this new feature and improving and testing `squash` module.
> 
> Great to see moving code from `dracut.sh` to individual dracut modules to improve modularity. Thanks !
> 
> One observation I have (independently from the `erofs` feature addition) that the new `squash_install` functions (old `installpost` function) makes certain assumptions about `busybox` build flags. It assumes for example that that `busybox` was compiled with `switch_root` support.
> 
> `busybox` can be installed using the `busybox` [dracut module](https://github.com/dracut-ng/dracut-ng/blob/main/modules.d/99busybox/module-setup.sh), which has support for mixing `busybox` applets and `coreutils` implementations when needed (so that `switch_root` could be provided by the `coreutils` implementation) .
> 
> I was wondering if you would consider installing `busybox` calling `module_install` (similarly as `squash_get_handler`) gets installed.

TBH, that's something I simply took over from the old implementation. But you are right using the busybox module should be more robust and prevents unnecessary duplicate code. I'll give it a try.

prudo1 commented:
fixed shfmt and shellcheck findings as well as commit messages. No functional change.

Conan-Kudo reviewed (approved):
This looks great and should allow us to scale to more backends (btrfs, wim, etc.).

prudo1 commented:
changes in new version:
- rebased to latest main
- added proper support to lsinitrd
- added support to dracut-initramfs-restore
  Note: I have no clue how to test this code. But the change is straight forward and should (tm) work
- added fix for missing directories in 99squash
- make use of 99busybox instead of manually installing it, as suggested by @LaszloGombos

LaszloGombos commented:
Thank you very much @prudo1 

> This leads to a dependency cycle as we want to allow users both to choose 99squash, when the exact filesystem doesn't matter

This usually gets resolved by introducing a dracut module that is responsible for making a decision for which dracut module is actually backing/implementing the functionality.  `dbus`, `network` and `shell-interpreter` are such modules.

The changes introduced have a side-effect of not showing some of the warning about missing transitive dependencies while building the initrd. As an example the following warning messages are now gone (can be observed on CI runs) for Alpine. That said the lack of warnings does not impact initrd generation.

```
dracut[E]: Module 'rngd' depends on module 'systemd', which can't be installed
dracut[I]: 62bluetooth: Could not find any command of '/usr/lib/bluetooth/bluetoothd /usr/libexec/bluetooth/bluetoothd'!
dracut[I]: 90systemd-cryptsetup: Could not find any command of '/systemd-cryptsetup'!
dracut[I]: 95nfs: Could not find any command of 'rpcbind portmap'!
```

LaszloGombos reviewed (approved):
LGTM. 

Thank you !

jozzsi commented:
> The changes introduced side-effect ... does not impact initrd generation.

Working on some of the PRs I noticed that this PR seems to regress initrd generation as now even dracut modules where dependencies are _not_ met are included in initrd (e.g. rngd for Alpine). I think this regression might need to be addressed.
